stapler/stapler.go

```diff
@@ -0,0 +1,529 @@
+// package stapler implements OCSP stapling feature described here: http://en.wikipedia.org/wiki/OCSP_stapling
+// stapler provides implementation that caches the staple on the first request and periodically updates the stapler
+package stapler
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"github.com/mailgun/vulcand/Godeps/_workspace/src/golang.org/x/crypto/ocsp"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/mailgun/vulcand/engine"
+
+	"github.com/mailgun/vulcand/Godeps/_workspace/src/github.com/mailgun/log"
+	"github.com/mailgun/vulcand/Godeps/_workspace/src/github.com/mailgun/timetools"
+)
+
+// Stapler is an interface for the OCSP staple cache implementations
+type Stapler interface {
+	// HasHost returns true if Stapler holds the response in cache
+	HasHost(host engine.HostKey) bool
+	// StapleHost returns the relevant StapleResponse, or error in case if response is unavailable
+	StapleHost(host *engine.Host) (*StapleResponse, error)
+	// DeleteHost deletes any OCSP data associated with the host entry
+	DeleteHost(host engine.HostKey)
+	// Subscribe subscribes the channel to the series of OCSP updates
+	Subscribe(chan *StapleUpdated, chan struct{})
+	Close()
+}
+
+// StaplerOption is used for optional parameters for the New function
+type StaplerOption func(s *stapler) error
+
+// Clock is an optional argument to the New function, by default the system clock is used
+func Clock(clock timetools.TimeProvider) StaplerOption {
+	return func(s *stapler) error {
+		s.clock = clock
+		return nil
+	}
+}
+
+// New returns a new instance of in-memory Staple resolver and cache
+func New(opts ...StaplerOption) (Stapler, error) {
+	s := &stapler{
+		v:           make(map[string]*hostStapler),
+		mtx:         &sync.Mutex{},
+		eventsC:     make(chan *stapleFetched),
+		closeC:      make(chan struct{}),
+		subscribers: make(map[int32]chan *StapleUpdated),
+		kickC:       make(chan bool),
+		discardC:    nil,
+		// setup transport more aggressive timeouts for OCSP staple responses
+		client: &http.Client{
+			Transport: &http.Transport{
+				Dial: (&net.Dialer{
+					Timeout:   10 * time.Second,
+					KeepAlive: 5 * time.Second,
+				}).Dial,
+				TLSHandshakeTimeout: 5 * time.Second,
+			},
+		},
+	}
+	for _, o := range opts {
+		if err := o(s); err != nil {
+			return nil, err
+		}
+	}
+	if s.clock == nil {
+		s.clock = &timetools.RealTime{}
+	}
+	go s.fanOut()
+	return s, nil
+}
+
+type stapler struct {
+	v       map[string]*hostStapler
+	mtx     *sync.Mutex
```

Comment:
If you make this field anonymous then (change it to `sync.Mutex`) you can just do:

```
s.Lock()
defer s.Unlock()
```